---
title: "Lab Retreat Talk"
author: "Flic Anderson"
date: "06/11/2021"
output: 
  ioslides_presentation
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.12.2/jquery.min.js"></script>

<script>
    $(document).ready(function() {
      $('slide:not(.title-slide, .backdrop, .segue)').append('<footer label=\"bit.ly/#######"></footer>');    
    })
</script>

<style>
  footer:after {
    content: attr(label);
    font-size: 12pt;
    position: absolute;
    bottom: 20px;
    left: 100px;
    line-height: 1.9;
  }
</style>

<style type="text/css">
slides > slide:not(.nobackground):after {
  content: '';
}
</style>

<div class="notes">

Bio: 

Flic Anderson is a Research Assistant in Bioinformatics in the Wallace Lab (within the Institute for Cell Biology) at the University of Edinburgh. The Wallace Lab works to understand how translation works within cells (particularly of fungi) and what factors control and regulate protein synthesis. 

Flic is part of the development team for the 'riboviz' open source software package for processing and analysis of ribosome profiling data, working with colleagues from EPCC (University of Edinburgh) to make the codebase more robust & sustainable.   

With a background in Ecology and Botany (previously a Research Assistant at the Royal Botanic Garden Edinburgh), Flic has worked with a wide range of types of biological data using different software tools and programming languages (particularly R) for the last 8 years, and found herself drawn towards bioinformatics and research software engineering. 

Flic is a member of the Edinburgh Carpentries and a certified Carpentries instructor in foundational coding and data science skills for researchers. 


Title: 
#########


Abstract:  

'riboviz' is an open source software package for processing and analysing ribosome profiling (RNA sequence) data.

Including a short primer on what Workflow Management Systems such as Nextflow 'do' and why they're useful for complex data pipelines like riboviz

</div>


## HOW TO (RE-)USE THIS MATERIAL

This is a `.html` presentation created in `R Markdown` with `ioslides`. 

(It's been written in a [.Rmd](http://rmarkdown.rstudio.com) file, and I generated .html slides by 'knitting' it in Rstudio.)

You can check out the code used to make these slides at the [Talk repo](https://github.com/FlicAnderson/LabRetreat_20211108) on Github, and adapt it for your own presentations if you like - there's a MIT Licence on the repo, which means: 

*"Basically, you can do whatever you want as long as you include the original copyright and license notice in any copy of the software/source."*

Source: [tl;drLegal](https://tldrlegal.com/license/mit-license)

# TALK TITLE HERE 

## Me!

<div style="float: left; width: 30%;">
```{r, out.width = "300"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/EdinbR-20210519/main/images/Me.jpg")
```  

[GitHub](https://github.com/FlicAnderson), [Twitter](https://twitter.com/Flic_Anderson), [LinkedIn](https://www.linkedin.com/in/flicanderson/)
</div>

<div style="float: right; width: 55%;">
### Felicity 'Flic' Anderson  

**Research Assistant in Bioinformatics** for 2.5 years at the [Wallace Lab](https://ewallace.github.io/), Institute for Cell Biology, University of Edinburgh  

* **developer** on riboviz project  - _my first project coding in a team_ 
* co-supervised 3 undergrad & 1 masters student project focussing on using riboviz
* certified **Carpentries instructor** (foundational data, software & coding courses)
* background mainly in **R**, some SQL & Bash, new-ish to Python
</div>

<div class="notes"> 

https://raw.githubusercontent.com/FlicAnderson/EdinbR-20210519/main/images/Me.jpg

https://ewallace.github.io/ - Wallace lab website

https://edcarp.github.io/ - Edinburgh Carpentries, local chapter of The Carpentries organisation (https://carpentries.org/)
</div>

# Outline: 

## Outline: 

* riboviz & ribosome profiling
* Workflow Management Systems
* Nextflow in a nutshell  


## Riboviz

**Riboviz is open source software which processes & analyses ribosome profiling (sequencing) data**  

*Developed in collaborative project BBSRC-NSF funded project*

Ribosome profiling data helps unlock details of **translation**:   
_how the cell creates proteins from genetic code (DNA, RNA)._ 

riboviz helps researchers check **quality control** & understand the **characteristics** of the experimental data - ie. _do different experimental conditions change how the cell translates proteins from genes?_

<font size="2">riboviz at github: https://github.com/riboviz/riboviz</font>

<div class="notes">
...
</div>


## Riboviz

Developing/improving riboviz software = **more researcher time** for biological questions

_More biological code, less programming code!_

The project's aims are to apply software engineering techniques to create more robust & reliable code:

riboviz at github: https://github.com/riboviz/riboviz

<div class="notes">

more researcher time: rather than tinkering with pipelines & bespoke analysis code…  

Robust & reliable code gives more confidence in results

</div>


## Riboviz Workflow: Inputs

<div class="centered">
```{r, out.width = "800"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/riboviz-workflow_inputs.png")
``` 
</div>

<div class="notes"> 

Additional organism-specific data files: ie the pipeline can run without them, but will not produce specific plots without them. 

</div>


## Riboviz Workflow: Analysis

<div class="centered">
```{r, out.width = "800"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/riboviz-workflow_analysis.png")
``` 
</div>

<div class="notes">

Analyse & quantify data:   
Create outputs (including for quality-control, further analysis)

One main R script which generates these plots & various analysis data; this is run for each sample, outputting .tsv and .pdf files into a folder for each sample within the overall dataset output folder.

</div>


## Riboviz Workflow: .PDF Outputs

<div class="centered">
```{r, out.width = "800"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/riboviz-workflow_outputs.png")
``` 
</div>  


<div class="notes">

Original outputs: .tsv data files and .pdf plot images for each sample. 

Dataset shown: B-Sc_2012: Saccharomyces cerevisiae yeast RPF samples at a range of meiotic phases from: High-resolution View of the Yeast Meiotic Program Revealed by Ribosome Profiling, Brar et al. 2012. https://doi.org/10.1126/science.1215110

Files present: ‘all’ - not all organisms have the same amount of data availale, and so information on tRNAs for example isn't available for all species, or isn't currently in a riboviz friendly format.  

This means we have some 'conditional' outputs. Not all taxa will have all of these outputs, which is something we've had to handle in the code. 

Yaml config file available at https://github.com/riboviz/example-datasets/blob/master/fungi/saccharomyces/Brar_2012_Meiosis_RPF_6-samples_CDS_w_250utrs_config.yaml

This was a dataset selected by Ania Kurowska (undergrad project in the Wallace lab, 2020)

</div>


## Riboviz Workflow: + Visualization

<div class="centered">
```{r, out.width = "800"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/riboviz-workflow_visualization.png")
``` 
</div>

<div class="notes">

As part of making riboviz more user-friendly, wanted to generate a per-sample .html report based on the summary data and pdf outputs already generated by riboviz

</div>  

# Workflow Management Systems (WfMS) & Nextflow

## Workflow Management Systems 

Workflow Management Systems (WfMS) provide infrastructure to set up, run & monitor a defined set of tasks (processes). 

 * **run time management** (the infrastructure supporting task management & execution) & **parallelism**

 * **portability / reproducibility** (same results, different machines)  

 * **faster troubleshooting** (with re-run options & log files)  
 
 * **easily shareable workflow scripts**  
 
Some examples are: Snakemake, Toil, cwltool and **Nextflow**


<div class="notes">   

* can easily 'containerise' code & dependencies for easier deployment in different computing environments 

* the same workflow can be run in different computing infrastructure such as HPC or local machine  

* the WfMS handles the different tasks and manages their execution - instead of running 'in order', tasks can be split up and parallelised to run at the same time; error handling & logging generated by system

* having the 'admin' handled by the WfMS means the same results on re-run, regardless of platform; specification is clearer  

* can share your workflow more easily as it includes specification details as well as describing the workflow  

* many WfMS let you re-run certain steps of a workflow if something has changed or broken, without needing to re-run ALL parts ('re-entrancy')  

My colleagues involved on riboviz wrote a paper on how we selected the right workflow management system for the project: https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1008622  

These tasks might depend on each other, for example as in a data processing pipeline (where the output of one stage becomes the input of the next step) or allow parts of the pipeline to be re-run without needing to start from the beginning. 

A Workflow Management System lets you join up all the tasks into one workflow, which you can run with one command. No chopping & changing between software, languages, or different environments!


</div>

## Why not just use a script?

'riboviz' _was_ previously controlled by a **python script** which ran other scripts (bash, python, R)

_More complex pipeline -> more admin-related development time!_ 

WfMS handle the admin-related tasks better (check **files exist**, create **log files**, queue & **run tasks efficiently**)

_Rather than creating **bespoke code**, we can use existing features_ - following scientific computing best practice: _"Re-use code instead of re-writing it"_  

<font size=2>Wilson et al. (2014) [Best Practices for Scientific Computing](https://doi.org/10.1371/journal.pbio.1001745). PLoS Biol 12(1): e1001745. </font>

<div class="notes">

More details on riboviz and requirements for a WfMS in the paper: https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1008622  

'riboviz' was previously run by a python script. Running this script then carried out all the data processing and analysis in bash, python and R. 

But as we added more functionality, extra tools and features, and it became more complex, we spent more time writing features around running the steps of the pipeline and less actual biological feature content!  

WfMS handle more of the process-related admin such as ensuring task inputs exist, creating log files, and queuing up the tasks to run in the right order and in an efficient manner according to the computing infrastructure. 

This latter point is key: in a script, tasks will happen sequentially based on their position. Having a WfMS splitting up these tasks and running them as soon as they're 'ready' makes it much more efficient (important if you're running big datasets through complex pipelines which can take hours!)

Think of a script which runs lots of separate functions to do different data analysis tasks: you can execute the whole script, and all the commands inside are run. But, often workflow management systems allow several of those commands to be run at the same time, rather than waiting until their 'turn' in the script. 

</div>


## Nextflow

Lots of Workflow Management System Benefits, including: 

* tasks split up efficiently & run based on **availability of inputs** - no waiting for unrelated tasks!  
* **resume option** - lets you re-run the pipeline from the last successful point if it stops ("re-entrancy")  
* **simple syntax** so you can re-use existing scripts
* you can use **containerisation** (such as docker/singularity) and **conda** to bundle and/or manage your software

<font size="2">Jackson, Kavoussanakis & Wallace. (2021). [Using prototyping to choose a bioinformatics workflow management system](https://doi.org/10.1371/journal.pcbi.1008622)</font>  

<div class="notes">  

https://www.nextflow.io/

Find out more: Nextflow Carpentries style workshop (alpha!) lesson materials under development https://carpentries-incubator.github.io/workflows-nextflow/

Nextflow connects 'processes' (the steps or tasks) according to their outputs and inputs to other processes. 
 
Nextflow combines a **runtime environment** (the stuff that runs other stuff) and a **domain specific language** (a language which is specific to a particular type of task or activity, and not really intended for use outside it's area).

DSL: html is a DSL for web pages for example, SQL for databases. 


Answer: 
* what is it
* why do we use it?
* what was the reasoning of switching to this from python/bash?

</div>


## Thanks/Acknowledgements

riboviz is a BBSRC-NSF funded collaborative project: 


* **The Wallace Lab** - *University of Edinburgh*:  
Edward Wallace, Emma MacKenzie, Sophie Winterbourne + past students (Ania Kurowska, Siyin Xue)  
* **The Shah Lab** -  *Rutgers University*:  
Premal Shah, John Favate, Alex Cope, Peter Tilton  
* **The Lareau Lab** - *University of California, Berkeley*:  
Liana Lareau, Amanda Mok, Vikram Shivakumar  
* **EPCC** - *University of Edinburgh*:  
Kostas Kavousannakis, Mike Jackson, Junchen Liu  


<div class="notes">
... slide notes ...
</div>


# Questions?


<div class="notes">
... slide notes ...
</div> 
